### PR TITLE
Add Promise API

### PR DESCRIPTION
**Pull request by wbinnssmith**
Adding Promise support as we discussed [here](https://github.com/sindresorhus/del/pull/29#issuecomment-136175242).

I was a little unsure with this since I think it could benefit from supporting both callbacks and promises, but it looks like 99% of the usage of this module is paired with `del` as gulp recommends.

Also added xo to keep me sane and since you're moving your stuff towards it :stuck_out_tongue:

### CODE REVIEW

**Comment by sindresorhus** [comment]
Can you update to readme to mention requiring a promise returning function? Maybe also add another example that uses a promise explicitly. E.g:

```js
gulp.task('delete', function () {
    return gulp.src('app/*')
        .pipe(stripDebug())
        .pipe(vinylPaths(function (paths) {
            console.log('Paths:', paths);
            return Promise.resolve();
        ));
});
```

File: index.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ module.exports = function (userCb) {
 		this.paths.push(file.path);
 
 		if (userCb) {
-			userCb(file.path, function () {
+			userCb(file.path).then(function () {
```
Comment (wbinnssmith) [inline index.js]: There's potential here for an unhandled promise rejection here. Is a rejected promise from the user cause enough to end the stream? Maybe we could warn and continue?

Comment (sindresorhus) [inline index.js]: Yes, rejected promise should err the stream:

```js
userCb(file.path).then(function () { ... }).catch(cb);
```

Right?

**Comment by wbinnssmith** [comment]
re: the `delete2` example in the readme, the best I can come up with as an analog is this:

```js
// or if you need to use the paths after the pipeline
gulp.task('delete2', function () {
	return new Promise(function (resolve, reject) {
		var vp = vinylPaths();

		gulp.src('app/*')
			.pipe(vp)
			.pipe(gulp.dest('dist'))
			.on('end', function () {
				del(vp.paths).then(resolve).catch(reject);
			});
	})
});
```

...which seems a bit less elegant than what we had with callbacks. It's not *that* bad though, unless you have another suggestion?

**Comment by sindresorhus** [comment]
I was talking about this example:

```js
gulp.task('delete', function () {
    return gulp.src('app/*')
        .pipe(stripDebug())
        .pipe(vinylPaths(function (paths) {
            console.log('Paths:', paths);
            return Promise.resolve();
        ));
});
```

Do you think it's good enough as a second example for the main method of using it? I just want to show it used without directly just passing in `del` as people might not be aware `del` returns a promise.

Good point with your example though. We need to update the existing second example with promises.

File: index.js (hunk 1 above)
Comment (wbinnssmith) [inline index.js]: yep!

**Comment by wbinnssmith** [comment]
All right, I think this is in a much better place now. Let me know what you think.

**Comment by sindresorhus** [comment]
Looks good! :beers:

**Pull request closed without merging**